### PR TITLE
Introduce Rule.Context

### PR DESCRIPTION
**Pull request by kokosing**
Introduce Rule.Context

Motivation:
 - when implementing a rule it is easier to pass rule context parameters
 to other methods.
 - less boilerplate in rule implementations
 - easier to modify Rule#apply method parameters

### CODE REVIEW

**Comment by kokosing** [comment]
@martint Are you OK with such refactor?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java [hunk 1]
```diff
@@ -111,7 +111,7 @@ public static boolean isOriginalOrder(List<Integer> joinOrder)
 
         PriorityQueue<PlanNode> nodesToVisit = new PriorityQueue<>(
                 graph.size(),
-                (Comparator<PlanNode>) (node1, node2) -> priorities.get(node1.getId()).compareTo(priorities.get(node2.getId())));
+                comparing(node1 -> priorities.get(node1.getId())));
```
Comment (ArturGajowy) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/EliminateCrossJoins.java]: nit: `node1` -> `node`

**Comment by ArturGajowy** [comment]
I'd consider exposing delegating methods in the context itself, like
`Context#nextPlanNodeId()` or `Context#resolve`. Probably just implementing all the interfaces composing the Context would make the interfaces' methods look out of context :wink:

**Comment by kokosing** [comment]
@ArturGajowy 
>I'd consider exposing delegating methods in the context itself, like
Context#nextPlanNodeId() or Context#resolve. Probably just implementing all the interfaces composing the Context would make the interfaces' methods look out of contex

I tried that, and then when I looked at the `Context` API I was scarred. It had all the responsibilities. I think that current way is better.

**Comment by kokosing** [comment]
@ArturGajowy 

> I like the idea - allows for less noise in the signature and for easy check which Rule uses which facilities (now it's just a usages search on Context methods)

It looks like just read a good article how to give a good review ;)

**Comment by ArturGajowy** [comment]
> It looks like just read a good article how to give a good review ;)

I was giving positive feedback long before reading that article ;)

**Comment by kokosing** [comment]
@sopel39 
This may relates to https://github.com/prestodb/presto/issues/7169 (see point 2, where a context could be a plan alternatives consumer)

File: presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java [hunk 2]
```diff
@@ -170,7 +170,7 @@ private boolean exploreChildren(int group, Context context)
         return progress;
     }
 
-    private static class Context
+    private static class Context implements Rule.Context
```
Comment (findepi) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/IterativeOptimizer.java]: fmt

**Comment by martint** [comment]
> @martint Are you OK with such refactor?

Yeah, this is fine. I expected we'd have to do it at some point.

**Pull request merged**